field/src/integers.rs

```diff
@@ -0,0 +1,452 @@
+/// impl QuotientMap<i128> for Mersenne31 {
+///     /// Convert a given `i128` integer into an element of the `Mersenne31` field.
+///     ///
+///     /// This checks the sign and then makes use of the equivalent method for unsigned integers.
+///     /// This should be avoided in performance critical locations.
+///     #[inline]
+///     fn from_int(int: i128) -> Mersenne31 {
+///         if int >= 0 {
+///             Self::from_int(int as u128)
+///         } else {
+///            -Self::from_int(-int as u128)
+///         }
+///     }
+///
+///     /// Convert a given `i128` integer into an element of the `Mersenne31` field.
+///     ///
+///     /// Returns `None` if the input does not lie in the range: `[-2^30, 2^30]`.
+///     #[inline]
+///     fn from_canonical_checked(int: i128) -> Option<Mersenne31> {
+///         // We just check that int fits into an i32 now and then use the i32 method.
+///         let int_small = TryInto::<i32>::try_into(int);
+///         if int_small.is_ok() {
+///             Self::from_canonical_checked(int_small.unwrap())
+///         } else {
+///             None
+///         }
+///     }
+///
+///     /// Convert a given `i128` integer into an element of the `Mersenne31` field.
+///     ///
+///     /// # Safety
+///     /// The input mut lie in the range:", `[1 - 2^31, 2^31 - 1]`.
+///     #[inline]
+///     unsafe fn from_canonical_unchecked(int: i128) -> Mersenne31 {
+///         Self::from_canonical_unchecked(int as i32)
+///     }
+/// }
+///```
+#[macro_export]
+macro_rules! quotient_map_large_iint {
+    ($field:ty, $field_size:ty, $checked_bounds:literal, $unchecked_bounds:literal, [$(($large_signed_int:ty, $large_int:ty)),*] ) => {
+        $(
+        impl QuotientMap<$large_signed_int> for $field {
+            #[doc = concat!("Convert a given `", stringify!($large_signed_int), "` integer into an element of the `", stringify!($field), "` field.
+                \n This checks the sign and then makes use of the equivalent method for unsigned integers. \n This should be avoided in performance critical locations.")]
+            #[inline]
+            fn from_int(int: $large_signed_int) -> $field {
+                if int >= 0 {
+                    Self::from_int(int as $large_int)
+                } else {
+                    -Self::from_int(-int as $large_int)
+                }
+            }
+
+            #[doc = concat!("Convert a given `", stringify!($large_int), "` integer into an element of the `", stringify!($field), "` field.
+                \n Returns `None` if the input does not lie in the range:", $checked_bounds, ".")]
+            #[inline]
+            fn from_canonical_checked(int: $large_signed_int) -> Option<$field> {
+                let int_small = TryInto::<$field_size>::try_into(int);
+                if int_small.is_ok() {
+                    Self::from_canonical_checked(int_small.unwrap())
+                } else {
+                    None
+                }
```

Comment:
Nit: Arguably cleaner/more idiomatic?
```suggestion
                int_small.map(Self::from_canonical_checked).ok()
```
Would need to update the comment above too.